service/history/shard/context_impl.go

```diff
@@ -2127,6 +2129,16 @@ func newContext(
 			false,
 		)
 	}
+
+	// We need to set lastUpdate time to "now" + "wait between shard updates time" -  "first update interval"
+	// This is done to make sure that first shard update` will happen around "first update interval".
+	// The idea is to allow queue to persist even in the case of (relativly) constantly
+	// moving shards between hosts.
+	// Note: it still may prevent queue from progressing if shard moving rate is too high
+	lastUpdated := shardContext.timeSource.Now()
+	lastUpdated = lastUpdated.Add(shardContext.config.ShardUpdateMinInterval())
+	lastUpdated = lastUpdated.Add(-1 * shardContext.config.ShardFirstUpdateInterval())
+	shardContext.lastUpdated = lastUpdated
```

Comment:
For testing, I think we can create a test shard context with fake time source (it's called eventTimeSource I think). Issue some update requests make sure there's no call to DB, then advance the time source to after the firstUpdateInterval make another update call, and assert the update does go to persistence.

Reply:
add test in next iter.